[export] fakify module state in nonstrict

suo opened the pull request:
Summary:
Previously, we were not fakifying module state explicitly in the nonstrict path.

This led to errors when modules were constructed under a fake mode, since the user-provided fake mode was clashing with the one that we had constructed internally to fakify the inputs.

This fixes things to use a single fake mode for everything.

As a side effect, this raised the question of how we ought to serialize state_dicts/constants that might be fake tensors. Naively calling torch.save understandably explodes—so this diff piggybacks on our infra for doing this on meta["val"]. Open to revising this, I'm low confidence that it's the best way to do it.

Test Plan: unit tests

Differential Revision: D53484942

tugsbayasgalan commented on torch/_export/non_strict_utils.py:
```diff
@@ -80,6 +80,16 @@ def fakify(
+) -> Dict[str, Union[torch.Tensor, torch.nn.Parameter]]:
+    faked_params_buffers = {}
+    for key, value in params_buffers.items():
+        faked_params_buffers[key] = fake_mode.from_tensor(value, static_shapes=True)
```
Should we be using the fake mode from the already fakified params and buffers instead of using the freshly created fake mode? I am not sure if it matters in practice tho.

suo committed:
[export] fakify module state in nonstrict (#119297)

Summary:

Previously, we were not fakifying module state explicitly in the nonstrict path.

This led to errors when modules were constructed under a fake mode, since the user-provided fake mode was clashing with the one that we had constructed internally to fakify the inputs.

This fixes things to use a single fake mode for everything.

As a side effect, this raised the question of how we ought to serialize state_dicts/constants that might be fake tensors. Naively calling torch.save understandably explodes—so this diff piggybacks on our infra for doing this on meta["val"]. Open to revising this, I'm low confidence that it's the best way to do it.

Test Plan: unit tests

Reviewed By: tugsbayasgalan

Differential Revision: D53484942
